Changes the customize_slug depending on simple / atomic.

cpapazoglou opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Fixes https://github.com/Automattic/jetpack/issues/18596

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Support changing the customizer link ($customize_slug) through a parameter in the `add_appearance_menu` so that simple / atomic class can set the correct link.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

**On an atomic site**
* Load this branch with Jetpack Beta
* Check that customizer URLs point to WPAdmin links and work

**On a jetpack site**
* Load this branch
* Enable `nav-unification` by adding `add_filter( 'jetpack_load_admin_menu_class', '__return_true' );` in a mu-plugin

**On a simple site**
* apply D56179 patch
* Check that customizer URLs point to WordPress.com links and work

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
<!-- Guidelines: https://github.com/Automattic/jetpack/blob/master/docs/writing-a-good-changelog-entry.md -->
* Avoid redirects for customizer links

getdave commented on projects/plugins/jetpack/modules/masterbar/admin-menu/class-atomic-admin-menu.php:
```diff
@@ -229,4 +229,15 @@ public function add_options_menu( $wp_admin = false ) {
+	 * @param string $customize_slug. Whether customizer links should point to Calypso or wp-admin.
+	 */
+	public function add_appearance_menu( $wp_admin = false, $customize_slug ) {
+		$customize_slug = 'customize.php';
+		parent::add_appearance_menu( $wp_admin, $customize_slug );
```
For the purposes of this PR I think this is ok. In general however, I think it points to a potential issue with the code being overly reliant on a class hierarchy.

If we were to decouple this and use traits to "mixin" the functionality required into each class we could call `add_appearance_menu` directly and pass the parameters once, allowing us to avoid this polymorphic approach.

I've spoken to @mmtr about this as well as a potential avenue for refactoring.

cpapazoglou replied:
I am not fan of this either. I will fix the notices and let's refactor the approach on a new PR just to allow @jeherve to cherry pick this on Jetpack 9.4

getdave reviewed (approved):
Thanks for jumping on this so quickly.

* Lets tidy up the code standards issues and warnings.
* One small nit to note about.

Otherwise this tested well for me on Simple and Atomic.

### Simple
<img width="497" alt="Screen Shot 2021-01-29 at 14 27 00" src="https://user-images.githubusercontent.com/444434/106286924-1aad6e00-623e-11eb-85f6-3e07219cf81f.png">

### Atomic
<img width="571" alt="Screen Shot 2021-01-29 at 14 26 28" src="https://user-images.githubusercontent.com/444434/106286930-1bde9b00-623e-11eb-88bc-a76b0a9982f7.png">

cpapazoglou commented on projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php:
```diff
@@ -364,12 +364,12 @@ function ( $parent_file ) use ( $menu_slug ) {
+	 * @param String $customize_slug Whether customizer links should point to Calypso or wp-admin.
 	 */
-	public function add_appearance_menu( $wp_admin = false ) {
+	public function add_appearance_menu( $wp_admin = false, $customize_slug ) {
```
> Declaring a required parameter after an optional one is deprecated since PHP 8.0. Parameter $wp_admin is optional, while parameter $customize_slug is required. (PHPCompatibility.FunctionDeclarations.RemovedOptionalBeforeRequiredParam.Deprecated)

I would prefer to leave this as is cause:
- refactoring all the calls will introduce some noise which would not let easy cherry picking
- this code WILL be refactored in follow up PR https://github.com/Automattic/jetpack/pull/18586

cpapazoglou replied:
It seems though, that tests do not pass. Let me think about it

cpapazoglou replied:
Fixed d570570

mmtr commented:
Perfect example of how we can split https://github.com/Automattic/jetpack/pull/18586 into smaller PRs 🙂 (fix here is also part of the many changes done in https://github.com/Automattic/jetpack/pull/18586).

jeherve reviewed (changes requested):
This doesn't seem to work for me in Jetpack:

![image](https://user-images.githubusercontent.com/426388/106303757-cbc20180-625a-11eb-82c6-7cda9de38f76.png)

cpapazoglou commented:
> This doesn't seem to work for me in Jetpack:
> 
> ![image](https://user-images.githubusercontent.com/426388/106303757-cbc20180-625a-11eb-82c6-7cda9de38f76.png)

Thanks @jeherve ! It seems that I have totally ignored enabling `nav-unification` on Jetpack without `define('IS_ATOMIC', true);`.. 😢

This fix 485226c works for me. Can you confim?

obenland commented on projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php:
```diff
@@ -364,12 +364,12 @@ function ( $parent_file ) use ( $menu_slug ) {
+	 * @param Bool   $wp_admin Optional. Whether links should point to Calypso or wp-admin. Default false (Calypso).
 	 */
-	public function add_appearance_menu( $wp_admin = false ) {
+	public function add_appearance_menu( $customize_slug, $wp_admin = false ) {
```
If you set `$customize_slug` as a class property defaulting to `customize.php`, you'd only need to update it in `Atomic_Admin_Menu`'s constructor. It would avoid this change and you wouldn't need to add new methods in the child classes.

cpapazoglou replied:
@obenland thanks for that suggestion! I have refactored the PR and it seems working for Simple, Atomic, Jetpack. I still have some problems with the tests though. I will need to look it further!

jeherve commented:
> This fix 485226c works for me. Can you confirm?

Yep, that did it for me.

obenland commented on projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php:
```diff
@@ -44,7 +51,8 @@ protected function __construct() {
 		add_action( 'rest_request_before_callbacks', array( $this, 'rest_api_init' ), 11 );
 
-		$this->domain = ( new Status() )->get_site_suffix();
+		$this->domain          = ( new Status() )->get_site_suffix();
+		$this->$customize_slug = 'customize.php';
```
Looks like there's a `$` too much.
```suggestion
		$this->domain         = ( new Status() )->get_site_suffix();
		$this->customize_slug = 'customize.php';
```

cpapazoglou replied:
WoW, that was too simple and obvious.. I guess it was late for me.. 😜

obenland commented on projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php:
```diff
@@ -364,12 +372,11 @@ function ( $parent_file ) use ( $menu_slug ) {
 	 * Adds Appearance menu.
 	 *
-	 * @param bool $wp_admin Optional. Whether links should point to Calypso or wp-admin. Default false (Calypso).
+	 * @param Bool   $wp_admin Optional. Whether links should point to Calypso or wp-admin. Default false (Calypso).
```
This can probably stay lowercase

obenland commented on projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php:
```diff
@@ -381,7 +388,7 @@ public function add_appearance_menu( $wp_admin = false ) {
 		add_menu_page( esc_attr__( 'Appearance', 'jetpack' ), __( 'Appearance', 'jetpack' ), $appearance_cap, $themes_slug, null, 'dashicons-admin-appearance', 60 );
 		add_submenu_page( $themes_slug, esc_attr__( 'Themes', 'jetpack' ), __( 'Themes', 'jetpack' ), 'switch_themes', $themes_slug, null, 5 );
-		add_submenu_page( $themes_slug, esc_attr__( 'Customize', 'jetpack' ), __( 'Customize', 'jetpack' ), 'customize', $customize_slug, null, 10 );
+		add_submenu_page( $themes_slug, esc_attr__( 'Customize', 'jetpack' ), __( 'Customize', 'jetpack' ), 'customize', $this->$customize_slug, null, 10 );
```
This should be `$this->customize_slug`. Looks like it's a copy/paste issue.
```suggestion
		add_submenu_page( $themes_slug, esc_attr__( 'Customize', 'jetpack' ), __( 'Customize', 'jetpack' ), 'customize', $this->customize_slug, null, 10 );
```

cpapazoglou replied:
> Looks like it's a copy/paste issue.

and that's why searching for `customize_slug` always returned the correct results.. The problem was the `$`...

obenland commented on projects/plugins/jetpack/tests/php/modules/masterbar/test-class-admin-menu.php:
```diff
@@ -60,6 +67,7 @@ class Test_Admin_Menu extends WP_UnitTestCase {
 	 */
 	public static function wpSetUpBeforeClass( $factory ) {
 		static::$domain  = ( new Status() )->get_site_suffix();
+		static::$customize_slug = 'customize.php';
```
This should not be needed, the class should set it correctly itself.

obenland commented on projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php:
```diff
@@ -34,6 +34,13 @@ class Admin_Menu {
+	 *
+	 * @var string
+	 */
+	protected $customize_slug;
```
If you set the default here, it doesn't need to be set in the constructor
```suggestion
	protected $customize_slug = 'customize.php';
```

cpapazoglou commented:
@obenland I addressed all feedback and now tests are passing 🎉!

@getdave I refactored this PR per obenland recommendation. I tested it again in:
- Simple
- Atomic
- Jetpack

could you please give another ✅ ?

mmtr commented on projects/plugins/jetpack/modules/masterbar/admin-menu/class-wpcom-admin-menu.php:
```diff
@@ -14,14 +14,24 @@
  */
 class WPcom_Admin_Menu extends Admin_Menu {
 
+	/**
+	 * WPcom_Admin_Menu constructor.
+	 */
+	protected function __construct() {
+		parent::__construct();
+
+		$this->customize_slug = 'https://wordpress.com/customize/' . $this->domain;
+	}
```
Can the property be overriwritten out of the constructor? That would simplify the code, but TBH I don't know if this is possible in PHP:

```suggestion
	protected $customize_slug = 'https://wordpress.com/customize/' . $this->domain;
```

mmtr replied:
Scratch that. I think it's not possible, because we cannot set a default variable which depends on another var like `$this->domain`

mmtr commented on projects/plugins/jetpack/tests/php/modules/masterbar/test-class-admin-menu.php:
```diff
@@ -39,6 +39,13 @@ class Test_Admin_Menu extends WP_UnitTestCase {
+	 *
+	 * @var string
+	 */
+	public static $customize_slug;
```
Is this needed?

mmtr commented on projects/plugins/jetpack/tests/php/modules/masterbar/test-class-atomic-admin-menu.php:
```diff
@@ -40,6 +40,13 @@ class Test_Atomic_Admin_Menu extends WP_UnitTestCase {
+	 *
+	 * @var string
+	 */
+	public static $customize_slug;
```
I think this can be removed too.

mmtr commented on projects/plugins/jetpack/tests/php/modules/masterbar/test-class-wpcom-admin-menu.php:
```diff
@@ -40,6 +40,13 @@ class Test_WPcom_Admin_Menu extends WP_UnitTestCase {
+	 *
+	 * @var string
+	 */
+	public static $customize_slug;
```
And this one as well

mmtr reviewed (dismissed):
This tests well for me. I left some minor comments regarding unused variables.

cpapazoglou commented:
Thanks @mmtr! Removed the unused vars caa6409.

jeherve reviewed (approved):
This tests well for me. 🚢

jeherve commented:
Cherry-picked to `jetpack/branch-9.4` in 1baa47c28f9ed6588c83355779785450a8386235. I'll let you handle the wpcom matching diff.

cpapazoglou commented:
Deployed r220316-wpcom.
